### ISSUE TITLE
[BUG] Intermittent Issue where fixture is not returning unique data

### ISSUE THREAD

**Issue by costag1982** [microsoft/playwright#17963]
**Context:**
- Playwright Version: 1.26.1
- Operating System: Mac
- Node.js version: v16.15.0
- Browser: N/A
- Extra: MacBook Pro M1 Pro

## System:
 - OS: macOS 13.0
 - Memory: 576.33 MB / 32.00 GB
## Binaries:
 - Node: 16.15.0 - ~/.nvm/versions/node/v16.15.0/bin/node
 - Yarn: 1.22.19 - /opt/homebrew/bin/yarn
 - npm: 8.5.5 - ~/.nvm/versions/node/v16.15.0/bin/npm
## Languages:
 - Bash: 3.2.57 - /bin/bash

Help us help you! Put down a short code snippet that illustrates your bug and
that we can run and debug locally. For example:

**FIXTURE**

```javascript
import { test as base } from '@playwright/test';
import { faker } from '@faker-js/faker';

type Email = {
  email: string;
};

export const test = base.extend<{}, { email: Email }>({
  email: [async ({ browser }, use, workerInfo) => {
    const email = faker.internet.exampleEmail();
    await use({ email });
  }, { scope: 'worker' }],

});

export default test;
export const expect = test.expect;
```

**TEST**

```javascript
import test, { expect } from '../../fixtures/testFixture';

test.describe.configure({ mode: 'parallel' });

test('Test One', async ({email}) => {
  console.log(email.email);
});

test('Test Two', async ({email}) => {
  console.log(email.email);
});
```

**OUTPUT**

```
Running 2 tests using 2 workers
Mitchel_Senger89@example.org
·Mitchel_Senger89@example.org
·

  2 passed (1s)
➜  play-on-e2e npm run test

> play-on-e2e@1.0.0 test
> npx playwright test


Running 2 tests using 2 workers
Keeley_Veum68@example.org
Lois15@example.com
··
```

**Describe the bug**
This example is a simplified version of the real issue I am facing in my actual e2e test project. I am using faker to generate fake data and passing this through to the test. The tests are running in parallel when this issue occurs. As you can see from the output where I print the fake email address on the first run the emails are the same but on the 2nd run the emails are unique.

I believe that the desired behaviour is the emails being unique. I would expect this because they run on separate workers. At the very least I would expect the results to be consistent.

Please can you look into it for me?

Add any other details about the problem here.

The reason this is an issue in my real tests is firebase is returning a 400 for 1 of the 2 requests when creating a user as they have to have unique email addresses.

It passes on some occasions which explains the intermittent behaviour.

<img width="544" alt="fixture_test_ts_—_play-on-e2e" src="https://user-images.githubusercontent.com/12000247/194927124-010321db-ad10-496e-8260-6c53f9ceba0b.png">

Hope this helps.
Thanks

**Comment by pavelfeldman** [comment]
You are using 2 workers, they run independently in different OS processes. So they have different values of their worker-level fixtures, as expected. If you are looking for the unique email across the entire test run (for all workers), you can generate this email in the global setup and pass it to the workers using the environment variables.

**Comment by costag1982** [comment]
Thanks for getting back to me.

As you can see in my examples when I ran this twice.

On the first run the emails are the same which I wasn’t expecting. Then on the second run the emails were unique which is what I was expecting.

This is the intermittent behaviour I am referring to.

I have a single test file with 4 test scenarios. I want them to run in parallel and to do that the test data for each test needs to be unique.

If I doing this the wrong way please could you kindly help me?

I have looked at the global setup and played around with it last night and I could get it work but I would need to set up 4 unique email addresses and pass them in as 4 unique environment variables which seems more of a hack than a solution.

thanks
Costa

**Comment by pavelfeldman** [comment]
I'm not familiar with the way `@faker-js/faker` seeds its randomizer. If it does not, it would explain why two identical workers generate identical emails.

**Comment by costag1982** [comment]
Hi

I think you are on to something here.

I replaced faker with UUID

```javascript
import { test as base } from '@playwright/test';
import ObjectID from 'bson-objectid';

type Email = {
  email: string;
};

export const test = base.extend<{}, { email: Email }>({
  email: [async ({ browser }, use, workerInfo) => {
    const email = ObjectID().toHexString();
    await use({ email });
  }, { scope: 'worker' }],
});

export default test;
export const expect = test.expect;
```

and I get a unique value every time so maybe I can set a random seed.

Here is the output:

```
Running 2 tests using 2 workers
{ email: '6345c29f87a3547f2352f61e' }
{ email: '6345c29f624e337f24569567' }
```

I ran this numerous times and always got unique values.

Thanks for your quick replies my sanity is restored. :)

**Comment by costag1982** [comment]
I have found a solution where I generate a random number to pass into the faker seed

```javascript
import { test as base } from '@playwright/test';
import { faker } from '@faker-js/faker';

const randomIntFromInterval = (min: number, max: number): number => {
  return Math.floor(Math.random() * (max - min + 1) + min)
}

type Email = {
  email: string;
};

export const test = base.extend<{}, { email: Email }>({
  email: [async ({ browser }, use, workerInfo) => {
    const randomNum = randomIntFromInterval(0, 10000);
    faker.seed(randomNum);
    const email = faker.internet.exampleEmail();
    await use({ email });
  }, { scope: 'worker' }],

});

export default test;
export const expect = test.expect;
```

This seems to return unique emails now every time

```
> npx playwright test


Running 2 tests using 2 workers
{ email: 'Ray.Lind30@example.net' }
·{ email: 'Ashleigh79@example.com' }
·

  2 passed (3s)
➜  play-on-e2e 
```

Once again thanks for your help

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.

**Comment by pschroeder89** [comment]
Been pulling my hair out over this, thanks @costag1982 !
